src/setfit/modeling.py

```diff
@@ -41,13 +166,93 @@ def __init__(self, model_body=None, model_head=None, multi_target_strategy=None)
         super(SetFitModel, self).__init__()
         self.model_body = model_body
         self.model_head = model_head
+
         self.multi_target_strategy = multi_target_strategy
 
         self.model_original_state = copy.deepcopy(self.model_body.state_dict())
 
-    def fit(self, x_train, y_train):
-        embeddings = self.model_body.encode(x_train)
-        self.model_head.fit(embeddings, y_train)
+    def fit(
+        self,
+        x_train: List[str],
+        y_train: List[int],
+        num_epochs: Optional[int] = None,
+        batch_size: Optional[int] = None,
+        learning_rate: Optional[float] = None,
+        body_learning_rate: Optional[float] = None,
+        l2_weight: Optional[float] = None,
+    ):
+        if isinstance(self.model_head, nn.Module):  # train with pyTorch
+            self.model_body.train()
+            self.model_head.train()
+
+            dataloader = self._prepare_dataloader(x_train, y_train, batch_size)
+            criterion = self.model_head.get_loss_fn()
+            optimizer = self._prepare_optimizer(learning_rate, body_learning_rate, l2_weight)
+
+            for epoch_idx in range(num_epochs):
+                for batch in dataloader:
+                    features, labels = batch
+                    optimizer.zero_grad()
+
+                    outputs = self.model_body(features)
+                    predictions = self.model_head._forward(outputs["sentence_embedding"])
```

Comment:
Depending on the answer to my question about two `forward()` methods, it might also be worth implementing a `__call__` method in `SetFitHead` to simplify this step a bit

Reply:
Ya, I think by merging `forward()` and `_forward()` together as I answered in your question, I can simplify this step.

Reply:
Fixed!